impeller/docs/shader_optimization.md

```diff
@@ -0,0 +1,280 @@
+# Writing efficient shaders
+
+When it comes to optimizing shaders for a wide range of devices, there is no
+perfect strategy. The reality of different drivers written by different vendors
+targeting different hardware is that they will vary in behavior. Any attempt at
+optimizing against a specific driver will likely result in a performance loss
+for some other drivers that end users will run Flutter apps against.
+
+That being said, newer graphics devices have architectures that allow for both
+simpler shader compilation and better handling of traditionally slow shader
+code. In fact, straight forward "unoptimized" shader code filled with branches
+may significantly outperform the equivalent branchless optimized shader code
+when targeting newer GPU architectures.
```

Comment:
"when targeting newer architectures due to ..."

If there's a simple 1-phrase explanation or summary it would be good to fill that in here rather than leaving it a mystery. Absent that, maybe point to the PR or issue where we had that discussion.